### PR TITLE
feat(backend): implement mock execution endpoint for formsg

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

Users having to spam their own form with test submissions

## Solution

Implement a `Test Step with Mock Data` option where the execution data for the form connection will be mocked.

This is implemented based on this latest exhaustive list of fields
```json
{
    "form": {
        "admin": {
            "agency": {
                "shortName": "ogp",
                "fullName": "Open Government Products",
                "emailDomain": [
                    "open.gov.sg",
                    "isomer.gov.sg",
                    "form.gov.sg",
                    "pay.gov.sg",
                    "vault.gov.sg"
                ],
                "logo": "https://s3-ap-southeast-1.amazonaws.com/agency-logo.form.sg/ogp.png"
            }
        },
        "authType": "NIL",
        "endPage": {
            "title": "Thank you for filling out the form.",
            "buttonText": "Submit another response",
            "paymentTitle": "Thank you, your payment has been made successfully.",
            "paymentParagraph": "Your form has been submitted and payment has been made."
        },
        "form_fields": [
            {
                "ValidationOptions": {
                    "customVal": null,
                    "selectedValidation": null,
                    "_id": "65eeb7c5ff96e4057a32e31d"
                },
                "allowPrefill": false,
                "lockPrefill": false,
                "_id": "65eeb7c5ff96e4057a32e31c",
                "title": "Short answer",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "textfield",
                "globalId": "Q891nxNCBCJsnKpod6tj2a6UzDwf2KPn6Z4KIkT0SaJ"
            },
            {
                "ValidationOptions": {
                    "customVal": null,
                    "selectedValidation": null,
                    "_id": "65eeb7c7ef5d3fbd8a32fe4d"
                },
                "_id": "65eeb7c7ef5d3fbd8a32fe4c",
                "title": "Long answer",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "textarea",
                "globalId": "8GZ8nHzOG0LtDdjv9PzAD5Ik2GI3RwEDKEMRR9pwxzA"
            },
            {
                "fieldOptions": [
                    "Option 1",
                    "Option 2"
                ],
                "othersRadioButton": true,
                "_id": "65eeb84f1b0f02af609f3653",
                "title": "Radio",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "radiobutton",
                "globalId": "bRfKpPsc4R9Cm5NThojRLQV1A2LogSKVYlqCrp51yz3"
            },
            {
                "ValidationOptions": {
                    "customMax": null,
                    "customMin": null
                },
                "fieldOptions": [
                    "Option 1",
                    "Option 2"
                ],
                "othersRadioButton": true,
                "validateByValue": false,
                "_id": "65eeb8524d5246341e3d8549",
                "title": "Checkbox",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "checkbox",
                "globalId": "LxRvWwhnc6JDwwcKAGuZKoHYlg2hwiYVR1h8g8zBiKI"
            },
            {
                "fieldOptions": [
                    "Option 1",
                    "Option 2"
                ],
                "_id": "65eeb857ff96e4057a330d73",
                "title": "Dropdown",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "dropdown",
                "globalId": "vJOvNZ6Vj8XcgTa9FHRAs5EtIOY1pPZceLGtFdxykvT"
            },
            {
                "_id": "65eeb85a10a21b6809d6df73",
                "title": "Country/Region",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "country_region",
                "globalId": "katS8BJf3ALyqtSOeqI7FJVnBNMeJGlo3waVRvVQjyN"
            },
            {
                "_id": "65eeb8634d5246341e3d87c3",
                "title": "Heading",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "section",
                "globalId": "8t0qtkDPTQeOoKt13OgdmuJKKuAN11dho7iTAfsTdv1"
            },
            {
                "_id": "65eeb867e7798cdb74222062",
                "title": "Statement",
                "description": "fdsf",
                "required": true,
                "disabled": false,
                "fieldType": "statement",
                "globalId": "NAwMz4E6s5ZBro53hTm1fJefURwm4HUwd5uZ5z6zuS4"
            },
            {
                "_id": "65eeb869279a4e911983b4cc",
                "title": "Yes/No",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "yes_no",
                "globalId": "UJoBzBMiDzwzcjZkCecd4mfgEc9ZwbIHhUzgiTMNejo"
            },
            {
                "ratingOptions": {
                    "steps": 5,
                    "shape": "Star"
                },
                "_id": "65eeb8747f593ba17823ae51",
                "title": "Rating",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "rating",
                "globalId": "7CNZu9goS4Zq9Fse7cXsikmfAWuwCYyeV5PfcZBAxHQ"
            },
            {
                "autoReplyOptions": {
                    "hasAutoReply": false,
                    "autoReplySubject": "",
                    "autoReplySender": "",
                    "autoReplyMessage": "",
                    "includeFormSummary": false
                },
                "isVerifiable": false,
                "hasAllowedEmailDomains": false,
                "allowedEmailDomains": [],
                "_id": "65eeb878279a4e911983b93b",
                "title": "Email",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "email",
                "globalId": "NhKAwzHyw9I1vI2HsKeDZKNkhes6haBNB28uuLgVLDq"
            },
            {
                "allowIntlNumbers": false,
                "isVerifiable": false,
                "_id": "65eeb87b7f593ba17823b055",
                "title": "Mobile number",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "mobile",
                "globalId": "pxuO8uwMiHHbqL5dUaJ87DSYCh3KlOL3CYqOPAhaySz"
            },
            {
                "allowIntlNumbers": false,
                "_id": "65eeb884ff96e4057a3319a1",
                "title": "Home number",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "homeno",
                "globalId": "2F0YjeyOYhpm33C7A2TQHFBxhJH4YB4HQm2VsOxcZnU"
            },
            {
                "dateValidation": {
                    "customMinDate": null,
                    "customMaxDate": null,
                    "selectedDateValidation": null
                },
                "invalidDays": [],
                "_id": "65eeb887b46c9e0f7f3730fa",
                "title": "Date",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "date",
                "globalId": "FhtaBNSa4jcJl0xOIadC5O0tflj0TNMqwv3jk53z3lK"
            },
            {
                "url": "https://s3.ap-southeast-1.amazonaws.com/images.form.gov.sg/65eeb8981b0f02af609f4b4a-65eeb7b7ef5d3fbd8a32f422-1710143640296-headshot.jpeg",
                "fileMd5Hash": "vWmQb9uRePsk8En5kbZklg==",
                "name": "headshot.jpeg",
                "size": "0.03 MB",
                "_id": "65eeb898d31e7d0b5341f004",
                "title": "Image",
                "description": "fdsfds",
                "required": true,
                "disabled": false,
                "fieldType": "image",
                "globalId": "KA0phqkJmpJKwNnsGUf0ddUTjX0vDw6jtMJG5bQ5jh8"
            },
            {
                "minimumRows": 2,
                "maximumRows": null,
                "addMoreRows": false,
                "columns": [
                    {
                        "title": "Text,Field",
                        "required": true,
                        "_id": "65eeb8a4e7798cdb742235be",
                        "columnType": "dropdown",
                        "fieldOptions": [
                            "hello",
                            "wassup"
                        ]
                    },
                    {
                        "title": "Text ,Field",
                        "required": true,
                        "_id": "65eeb8a4e7798cdb742235bf",
                        "columnType": "textfield",
                        "ValidationOptions": {
                            "customVal": null,
                            "selectedValidation": null
                        }
                    }
                ],
                "_id": "65eeb8a4e7798cdb742235bd",
                "title": "Table",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "table",
                "globalId": "vW3d6HwpvZ8KFyhs9hNBTgfxdrVC9Gsmb37A3BCRsdA"
            },
            {
                "attachmentSize": "1",
                "_id": "65eeb8a9279a4e911983c47e",
                "title": "Attachment",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "attachment",
                "globalId": "nHbiJTSN2wVFGC3JTRy4cAFU8nGaC56H24okrgapyfD"
            },
            {
                "ValidationOptions": {
                    "LengthValidationOptions": {
                        "customVal": null,
                        "selectedLengthValidation": null
                    },
                    "RangeValidationOptions": {
                        "customMin": null,
                        "customMax": null
                    },
                    "selectedValidation": null,
                    "_id": "65eeb8ade8f4b92af5aa4e3f"
                },
                "_id": "65eeb8ade8f4b92af5aa4e3e",
                "title": "Number",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "number",
                "globalId": "p69CDpAJf5SjB8SVFieQmop9reTMKubyxzjKt0oxpas"
            },
            {
                "ValidationOptions": {
                    "customMax": null,
                    "customMin": null
                },
                "validateByValue": false,
                "_id": "65eeb8b1c89f9e178efc4027",
                "title": "Decimal",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "decimal",
                "globalId": "4gj9HgO9g2UJ4vroETx61PVcfjaS6CXNpO22mkUN93p"
            },
            {
                "_id": "65eeb8b46bd31a8217ce188e",
                "title": "NRIC/FIN",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "nric",
                "globalId": "cXr4OBTp2jbD9jq47wo5rBPdNIAfgphOOOTOfeBf0BH"
            },
            {
                "_id": "65eeb8b7aab9735268baaf8f",
                "title": "UEN",
                "description": "",
                "required": true,
                "disabled": false,
                "fieldType": "uen",
                "globalId": "j4u4JcTw4Mlbr4P6T7GUocXeDMDMSRKW6EOPM01XKVP"
            }
        ],
        "form_logics": [],
        "hasCaptcha": false,
        "hasIssueNotification": false,
        "startPage": {
            "colorTheme": "blue",
            "logo": {
                "state": "DEFAULT"
            }
        },
        "status": "PUBLIC",
        "title": "tesla",
        "_id": "65eeb7b7ef5d3fbd8a32f422",
        "responseMode": "encrypt",
        "payments_field": {
            "products_meta": {
                "multi_product": false
            },
            "enabled": false,
            "description": "",
            "name": "",
            "amount_cents": 0,
            "min_amount": 0,
            "max_amount": 0,
            "payment_type": "Products",
            "gst_enabled": true,
            "products": []
        },
        "publicKey": "Izb+dSMyp1wqBrgQjAe0774FMKkfCt/yN8RikDv6dTM="
    },
    "isIntranetUser": false
}
```

with their corresponding sample value from a submission

```json
{
  "fields": {
    "65eeb7c5ff96e4057a32e31c": {
      "order": 1,
      "answer": "fdsfsfdsf",
      "question": "Short answer",
      "fieldType": "textfield"
    },
    "65eeb7c7ef5d3fbd8a32fe4c": {
      "order": 2,
      "answer": "fdsfds\nfdsfds\nfdsf",
      "question": "Long answer",
      "fieldType": "textarea"
    },
    "65eeb84f1b0f02af609f3653": {
      "order": 3,
      "answer": "Others: henlo",
      "question": "Radio",
      "fieldType": "radiobutton"
    },
    "65eeb8524d5246341e3d8549": {
      "order": 4,
      "question": "Checkbox",
      "fieldType": "checkbox",
      "answerArray": [
        "Others: wassup"
      ]
    },
    "65eeb857ff96e4057a330d73": {
      "order": 5,
      "answer": "Option 1",
      "question": "Dropdown",
      "fieldType": "dropdown"
    },
    "65eeb85a10a21b6809d6df73": {
      "order": 6,
      "answer": "AFGHANISTAN",
      "question": "Country/Region",
      "fieldType": "country_region"
    },
    "65eeb8634d5246341e3d87c3": {
      "order": 7,
      "answer": "",
      "isHeader": true,
      "question": "Heading",
      "fieldType": "section"
    },
    "65eeb869279a4e911983b4cc": {
      "order": 8,
      "answer": "No",
      "question": "Yes/No",
      "fieldType": "yes_no"
    },
    "65eeb8747f593ba17823ae51": {
      "order": 9,
      "answer": "4",
      "question": "Rating",
      "fieldType": "rating"
    },
    "65eeb878279a4e911983b93b": {
      "order": 10,
      "answer": "fdsfds@gmail.com",
      "question": "Email",
      "fieldType": "email"
    },
    "65eeb87b7f593ba17823b055": {
      "order": 11,
      "answer": "+6588888888",
      "question": "Mobile number",
      "fieldType": "mobile"
    },
    "65eeb884ff96e4057a3319a1": {
      "order": 12,
      "answer": "+6566666666",
      "question": "Home number",
      "fieldType": "homeno"
    },
    "65eeb887b46c9e0f7f3730fa": {
      "order": 13,
      "answer": "01 Mar 2024",
      "question": "Date",
      "fieldType": "date"
    },
    "65eeb8a4e7798cdb742235bd": {
      "order": 14,
      "question": "Table (Text,Field, Text ,Field)",
      "fieldType": "table",
      "answerArray": [
        [
          "HELLO",
          "howdy"
        ],
        [
          "HELLO",
          "wassup"
        ]
      ]
    },
    "65eeb8a9279a4e911983c47e": {
      "order": 15,
      "answer": "s3:plumber-development-common-bucket:65ef381b01be1cbef919284c/65eeb8a9279a4e911983c47e/headshot.jpeg",
      "question": "Attachment",
      "fieldType": "attachment"
    },
    "65eeb8ade8f4b92af5aa4e3e": {
      "order": 16,
      "answer": "12",
      "question": "Number",
      "fieldType": "number"
    },
    "65eeb8b1c89f9e178efc4027": {
      "order": 17,
      "answer": "1.234",
      "question": "Decimal",
      "fieldType": "decimal"
    },
    "65eeb8b46bd31a8217ce188e": {
      "order": 18,
      "answer": "S1234567D",
      "question": "NRIC/FIN",
      "fieldType": "nric"
    },
    "65eeb8b7aab9735268baaf8f": {
      "order": 19,
      "answer": "T16GB0002G",
      "question": "UEN",
      "fieldType": "uen"
    }
  },
  "submissionId": "65ef381b01be1cbef919284c",
  "submissionTime": "2024-03-12T00:58:03.088+08:00"
}
```

## Tests

- [x] Query `mockExecuteFlow` with any formsg triggers

## Deploy Notes

N/A

### CODE REVIEW

**Review by liangyuanruo — APPROVED** [review]
code quality LGTM - inviting @ogp-weeloong to help take a 2nd look

File: packages/backend/src/apps/formsg/triggers/new-submission/get-mock-data.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  if (fieldType === 'checkbox') {
+    return 'answerArray'
+  }
+  return 'answer'
+}
+
+async function getMockData($: IGlobalVariable) {
+  try {
+    const { formId } = getFormDetailsFromGlobalVariable($)
+    const { data } = await $.http.get(`/v3/forms/${formId}`)
```
Comment (ogp-weeloong) [inline packages/backend/src/apps/formsg/triggers/new-submission/get-mock-data.ts:76]: We can use `https://form.gov.sg/api/v3/forms/${formId}/sample-submission` API to get sample submissions, this is the same as what they used in the zapier integration

**Review by ogp-weeloong — CHANGES_REQUESTED** [review]
Let's rely on FormSG's sample API instead of maintaining our own copy

File: packages/backend/src/apps/formsg/triggers/new-submission/get-mock-data.ts (hunk 1 above)
Comment (stanleynguyen) [reply packages/backend/src/apps/formsg/triggers/new-submission/get-mock-data.ts:76]: This endpoint is not ready yet me thinks. Was crashing their server jn when I tried it

**Comment by stanleynguyen** [comment]
Closing in favour of https://github.com/opengovsg/plumber/pull/506

**Pull request closed without merging**